Add blog post for random device support in Unikraft (#163)
unikraft/docs: content/en/blog/2022-12-03-random-device-support-in-unikraft.md

+++
title = "Random Device Support in Unikraft"
date = "2022-12-03T09:00:00+01:00"
author = "Sebastian Samoilescu, Andrei Mutu"
tags = ["blog"]
+++

### Unikraft & Security

Unikernels could be a double-edged sword when it comes to security.
Unikraft managed to implement security mechanisms such as Control Flow integrity, Address Sanitization and Stack Canary, increasing the degree of security within the unikernel.
However, the project still somewhat suffers in other security areas such as ASLR or Page Protection.
In this post we aim to discusses about the necessary primitives to improve and develop the unikernel security system were delivered.

### Randomness & Entropy

Randomness represents the apparent or actual lack of a pattern.
In theory, that denotes that a random series of events can not predict the following events by any meaning.
Entropy is a scientific concept that measures the degree of uncertainty and disorder in a variable or a system.

> A computer is a deterministic machine. Thus, the result will be the same whenever a programme runs on top of it.

Reviewer: ```suggestion
A computer is a deterministic machine.
Thus, the result will be the same whenever a program runs on top of it.
```

That means that a computer cannot produce pure random numbers through software.

### Random Number Generators

> Evenso a way to produce randomness is thought pure-random number generators algorithms or PRNG.

Reviewer: ```suggestion
Even so, a way to produce randomness is thought pure-random number generators algorithms or PRNG.
```

The primary advantage of PRNG is their efficiency, meaning that they can generate a large amount of information in a short time.
The efficiency property makes PRNG good enough to rely on them in most cases.
However, applications that require unpredictable information cannot use these algorithms as a base source.

Besides PRNG, some algorithms extract entropy from asynchronous physical events.
These algorithms are True Random Number Generators or TRNG.
They analyze physical phenomena like mouse movement, the time between keystrokes, disk operations, or network communication.
Additional to these sources, Intel has released a CPU feature that provides support for generating pure random numbers.

> RDRAND is the instruction for returning this information from Intel's chip.

Reviewer: ```suggestion
`RDRAND` is the instruction for returning this information from Intel's chip.
```

The random number generation component is one of the leading security modules in an
operating system.
Randomization is critical in many areas, such as Address Space Layout Randomization, secret key generation, password salting, and TCP sequence numbers.
As the previous paragraph pointed out, the random number generation component is necessary for the operating system's proper functioning and offering API to higher security layers.

### Goals

Before this project, Unikraft provided support just for generating pseudo-random numbers, and there were no devices to provide this information (e.g. urandom and random).
Therefore, there were some limitations regarding the security of the unikernel or running cryptography applications in Unikraft.
For this reason, the target of this project was to provide an enhanced random number generator component.

#### Improving The Curent State

The first goal of the project was to improve what already exists.

> For example, the ukswrand (the software random number generator) library contains various algorithms that generate random information based on a seed, which is critical to the security of this library.

Reviewer: ```suggestion
For example, the `ukswrand` (the software random number generator) library contains various algorithms that generate random information based on a seed, which is critical to the security of this library.
```

The project increased the entropy of the seed gathering information using the processor random number generator and unpredictable events within the unikernel.

The first step to increase the security and functionality of Unikraft was to use the module delivered by the biggest processor makers.
Hence, the Unikraft RNG can produce high-entropy random numbers by calling CPU instructions such as read-seed or read-random.

#### The Entropy Service Provider

The next major outcome that this project managed to deliver was an entropy service provider based on the Linux RNG.
This service is responsible for gathering noise from random events, storing this information in a secure environment, and finally delivering the random numbers at a higher rate and in a secure manner.
At this moment, the RNG tracks just the network and IRQ events in order to gather random noise.

#### Linux-like Devices

> And last but not least, the new random number generator provides two devices, /dev/urandom and /dev/random which are backed up by pure random number generators and a combination of true and pure random number generators in the case of the random device.

Reviewer: ```suggestion
And last but not least, the new random number generator provides two devices, `/dev/urandom` and `/dev/random` which are backed up by pure random number generators and a combination of true and pure random number generators in the case of the random device.
```
